Fix for "Content" panel messed up

paolo-sz opened the pull request:
New QT implementation of QHeaderView is rewritten in 5.10.1 and behaves differently on the property StretchLastSection (true by default).
Setting that property to false while creating the new (unused) QHeaderView for "Content" window solve the issue #8439.

thalieht commented:
I tried this and the columns are still reset after qbt restart. I also tried `StretchLastSection(true)` in another widget (peerlist) to see it's effect and there were no side-effects on the columns.

paolo-sz commented:
@thalieht 
Nice behavior...
Ok, let me elaborate a little more all the reasoning behind my changes.
I noticed that at almost every interaction I had with Content panel, the last column was stretched till the end of the panel itself for a couple of seconds and then it got back to its "normal" size. This is why I focused on the stretch property.

Once said that, the current pull request is the result of some other tries i did and I'm start thinking the real fix is something that was left behind in the process. I mean some saved config, surely not C code.

I will do some further tests when I will be back on my build environment but, in the meanwhile, I can describe the way I followed.

1. with the "original" code, I removed from qbittorrent.ini (C:\users\...) all the QByteArray lines related to Content panel (I do not remember exactly how that lines looks like but I'm pretty sure you know better than me what I'm talking about). This did not solve the issue.
2. I applied the same patch as on the pull request with an additional
    <code>header()->setStretchLastSection(false);</code>
    just before (this solved the issue)
3. I applied the current pull request code and the issue was still gone

I'm wondering if step 2 had left something saved in the config...

By the way, my environment is:
Win10 (64bit)
mingw-w64_x86_64
qbittorrent built as 64 bit executable

thalieht commented:
I confirm that `header()->setStretchLastSection(false)` fixes it.

thalieht commented:
Heh, and now `header()->setStretchLastSection(true)` has no effect.

Btw i'm not a dev, i answered because i think non of them moved to Qt 5.10.1 (because none took an interest in this issue) and to report my observation with your patch.

paolo-sz commented:
I think we are in the black magic fields (did you note the comment above the piece of code we are touching?)... :-)

Anyway, I'm not a developer too, but any comment is more than welcome and useful (like your, indeed).
Maybe they are not yet moved to 5.10.1 and so our comments are queued in low priority but, sooner or later, they'll must move to newer QTs and then any issue identified and "solved" will be useful...

Chocobo1 commented:
https://bugreports.qt.io/browse/QTBUG-66572

If it's really fixed with Qt 5.11 then I rather wait than adding workarounds.

>i think non of them moved to Qt 5.10.1 (because none took an interest in this issue) 

https://git.archlinux.org/svntogit/packages.git/commit/trunk?h=packages/qt5-base&id=50588f5f1cc38c5ea7621b14b24300abb031ebaf
I noticed the bug reports, but I wanted to wait and see if Qt will fix it and it seems they do.

paolo-sz commented:
Not really sure they fix the issue we see within qbittorrent...
I reverted my qbittorrent patch and I applied to QT 5.10.1 sources all the commit they did to QHeaderView stuff till now (https://github.com/qt/qtbase/commit/2950de715ec8c48511090dcd13e1c44ae23cb561 https://github.com/qt/qtbase/commit/c0e45ae851c96dfebdebfd1e85b7b7eee6540bd1 https://github.com/qt/qtbase/commit/9395f35cb18725995910531ca8b09f1d84efa96c) and the Content widget issue is still here... but there are very good chances I missed something...

In the meanwhile I fixed my commit with the temporary patch that seems to fix the issue and I will leave the pull request open waiting to see if it will be useful when QT 5.11 will be released.

Chocobo1 commented:
>and the Content widget issue is still here... but there are very good chances I missed something...

You could try out Qt 5.11 directly if you have the time, if it's still not fixed then there is good reason to accept this PR.

paolo-sz commented:
> You could try out Qt 5.11 directly if you have the time, if it's still not fixed then there is good reason to accept this PR.

I'm trying but the released full tar.xz is too "old" and does not include all the latest commits... while 5.11 branch is failing to compile...
I will keep you updated...

paolo-sz commented:
Finally I was able to compile with 5.11-beta1 (Feb. 28 release).
I tried the qbittorrent master branch my PR come from and the latest one... no chances, the issue is still happening

Chocobo1 commented:
@glassez 
Maybe we should consider merging this PR.

glassez commented:
>Maybe we should consider merging this PR.

I don't mind if it doesn't spoil the qBittorrent builds using previous Qt versions.
BTW, commit message is currently invalid and should be fixed.

paolo-sz commented:
> BTW, commit message is currently invalid and should be fixed.

I will fix it.

My only concern is that I do not deeply debug the reason of the issue and the reason why my PR is a fix.
The same exact code is used for other widgets but only the Content one is affected by the issue... my fear is that my PR is only a dirty workaround for something which is not yet totally clear...

glassez commented:
>BTW, commit message is currently invalid and should be fixed.

Still invalid. Please consider Coding Guidelines.

glassez commented:
Thank you!

Chocobo1 commented:
@paolo-sz 
Just FYI: seems upstream fixed it with https://code.qt.io/cgit/qt/qtbase.git/patch/?id=9395f35c

paolo-sz commented:
@glassez  @Chocobo1 
You are more than welcome!

>Just FYI: seems upstream fixed it with https://code.qt.io/cgit/qt/qtbase.git/patch/?id=9395f35c

Unfortunately I fear it is not the case. That commit was already included in QT-5.11-beta1 I compiled in my latest test and the issue was not gone. 
I will try to find some time (and mood...) to seriously debug what was happening there... I feel this issue and my dirty fix have all the chances to be a time bomb...

randyrandersn commented:
Just to let all know that this Qt 5.10.1 bug has been patched in Arch (qt5-base package - I think) and issue no longer occurs with qBt 4.0.4.
